### PR TITLE
if users.mutableUsers is false, require root password to be set

### LINKED ISSUE #3788: [nixos] Cannot set password after turning off users.mutableUsers
In NixOS, i have set `users.mutableUsers` to `false`, rebuilt the system, and on the next reboot i could only log in as **root** with empty password.  I tried to run `passwd`, but received the error message:

> passwd: cannot run `/nix/var/nix/profiles/default/bin/passwd': No such file or directory

### PR DESCRIPTION
**Pull request by domenkozar**
Running `nixos-install` with `users.mutableUsers = false;` currently doesn't work. Following PR requires root password to be set once immutable users are used.

Also fixes #3788

### CODE REVIEW

**Comment by k0ral** [comment]
Commit 9bc8bcb only fixes the issue at NixOS installation phase. This PR also included a fix in `nixos/modules/config/users-groups.nix`, why was that part omitted ?
I've just set `users.mutableUsers = false` on an already installed NixOS and now I can't log in anymore...

**Comment by iElectric** [comment]
Because not everyone wants password set to the root user.

**Comment by k0ral** [comment]
So nothing currently prevents you from shooting yourself in the foot, as I've just done.
I believe the NixOS manual, and the `mutableUsers` option's description should warn about this.

Would you consider reimplementing the safeguard from `nixos/modules/config/users-groups.nix` as an opt-out feature ? Locking oneself out of a machine is really annoying.

**Comment by edolstra** [comment]
We could warn if neither root nor any of the wheel users are accessible (where "accessible" means having no password or authorized SSH keys).

**Comment by iElectric** [comment]
Warning means the profile will be activated so it's too late :/

**Comment by k0ral** [comment]
That's why I'm suggesting an opt-out safeguard. People who want a password-less setup would disable it.

**Comment by nbp** [comment]
Ideally we should warn when building the configuration, with the assertions, but I guess we cannot rely on the passwd to be consistent over time, so I guess we should somehow fail during the activation script and suggest to rollback.

File: nixos/modules/installer/tools/nixos-install.sh [hunk 1]
```diff
@@ -249,8 +249,8 @@ NIXOS_INSTALL_GRUB=1 chroot $mountPoint \
 chroot $mountPoint /nix/var/nix/profiles/system/activate
 
 
-# Ask the user to set a root password.
-if [ -t 0 ] ; then
+# Ask the user to set a root password if it's not set already
+if grep -q '^root:!' $mountPoint/etc/shadow && [ -t 0 ] ; then
```
Comment (fread2281) [inline nixos/modules/installer/tools/nixos-install.sh]: IMO, this should be done anyways so nixos-install can be run as a nixos-rebuld from recovery without having to type root password every time. (or if it didn't finish installing)

**Comment by ktosiek** [comment]
Would a separate "mutable-password" option (+an assert with a descriptive explanation of the problem) be enough? So one has to either set a password through config for root/user in wheel, or declare that a password is set outside of it?

**Pull request closed without merging**